src/composables/nav.ts

```diff
@@ -21,3 +30,257 @@ export interface ActionItem {
+}
+
+/**
+ * useUrlSearchParams accepts an initial set of search query params
+ * in the shape of Record<string, URLParamValue> and tracks the values
+ * of each param on window.location.search.
+ *
+ * This is a one-way sync where non-nullish, non-falsey values are popped
+ * onto window.location.search as they are mutated.  The initial values are not read
+ * from an existing query string on window.location.search as each values
+ * type cannot be inferred safely.
+ *
+ * This composable ignores query params that may already exist on window.location.search
+ * leaving the intact as much as possible, since it's not possible to determine where they
+ * came from or what function they may serve.
+ *
+ * The return value is a ref of the initial set of params.
+ *
+ * Example Usage:
+ * interface SearchParams {
+ *    q: string
+ *    isActive: boolean
+ *    attributes: string[]
+ * }
+ *
+ * // NOTE: initial params values should come from the
+ * // server in the form of a page prop. Though some
+ * // exceptions may exist.
+ * const props = defineProps<{
+ *    initialParams: SearchParams
+ * }>
+ *
+ * const searchParams = useUrlSearchParams(props.initialParam)
+ *
+ * <BaseInput v-model="searchParams.q" label="Search" type="search" />
+ * <Checkbox v-model="searchParams.isActive" label="Is Active" />
+ * <MultiCheckboxes v-model="searchParams.attributes" label="Attributes" :options="attrOpts" />
+ *
+ * When initialized as {q: "", isActive: false, attributes: []}
+ * the query string is not immediately updated.  Updates
+ * begin once the Ref is mutated.
+ * /path
+ *
+ * When mutated to {q: "jimothy bobbitz", isActive: true, attributes: ['att-1', 'att-3']}
+ * /path?q=jimothy+bobbitz&isActive=true&attributes=att-1&attributes=att-3
+ *
+ * When mutated later back to {q: "", isActive: false, attributes: []}
+ * /path?q=&isActive=false
+ *
+ * @param initial any Record<string, URLParamValue> like interface
+ * @return Ref<T>
+ */
+export const useUrlSearchParams = <T>(initial: URLParams<T>) => {
+  const searchParams = ref(initial)
+
+  watch(
+    searchParams,
+    (p) => {
+      const params = new URLSearchParams(window.location.search)
+      for (const key in p) {
+        const val = p[key]
+
+        // avoid setting nullish values
+        if (val == null || val == undefined) {
+          params.delete(key)
+          continue
+        }
+
+        // append array types as multiple key:value pairs
+        if (Array.isArray(val)) {
+          // remove the current key(s) to avoid doubling up on existing values
+          params.delete(key)
+
+          val.forEach((val) => {
+            params.append(key, val.toString())
+          })
+          continue
+        }
+
+        // avoid setting falsey (zero) values until a truthy value was previously set
+        // otherwise all zero-valued keys will be populated with every mutation
+        if (!params.has(key) && !val) {
+          continue
+        }
```

Comment:
@profsmallpine check out this latest little tweak.  I think this strikes the balance we are looking for.